### PR TITLE
Nexus support move/delete for components in r hosted repositories

### PR DESCRIPTION
**Pull request by DreamVan**
[Jira issue NEXUS-21170](https://issues.sonatype.org/browse/NEXUS-21170)

Integration tests for current implementation are located in the NXRM main repository [IT PR](https://github.com/sonatype/nexus-internal/pull/4740)

### CODE REVIEW

File: nexus-repository-r-it/src/test/java/org/sonatype/nexus/plugins/r/internal/RClientFactory.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+import java.net.URI;
+
+import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.sonatype.nexus.testsuite.testsupport.NexusClientFactory;
+
+/**
+ * IT Client factory
+ * @since 1.next
```
Comment (sergeyonofreychuck) [inline nexus-repository-r-it/src/test/java/org/sonatype/nexus/plugins/r/internal/RClientFactory.java]: 1.1.next

File: nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RComponentDirector.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  }
+
+  @Override
+  public Component afterMove(final Component component, final Repository destination) {
+    EntityMetadata entityMetadata = component.getEntityMetadata();
+    assert entityMetadata != null;
+    EntityId componentId = entityMetadata.getId();
+
+    if (sourceRepositoryName != null) {
+      eventManager.post(new AssetUpdatedEvent(entityMetadata, sourceRepositoryName, componentId));
```
Comment (chernikovmaksym) [inline nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RComponentDirector.java]: If asset is created/deleted event should be posted automatically. Does it really necessary to manually post it here?

Comment (DreamVan) [reply nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RComponentDirector.java]: By staging we do not change an asset, we change components. To fire reindex metadata which are looking on assets state I need manually call this event.

File: nexus-repository-r-it/src/test/java/org/sonatype/nexus/plugins/r/internal/RClientFactory.java (hunk 1 above)
Comment (DreamVan) [reply nexus-repository-r-it/src/test/java/org/sonatype/nexus/plugins/r/internal/RClientFactory.java]: Done c78c8180544f6204b0063b1ba2c56a4abc81150b

File: nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RComponentDirector.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                              final Repository destination)
+  {
+    sourceRepositoryName = source.getName();
+    return component;
+  }
+
+  @Override
+  public Component afterMove(final Component component, final Repository destination) {
+    EntityMetadata entityMetadata = component.getEntityMetadata();
+    assert entityMetadata != null;
```
[the remark this answers is not in the document]
Comment (DreamVan) [reply nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RComponentDirector.java]: Done c78c8180544f6204b0063b1ba2c56a4abc81150b

File: nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RComponentDirector.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+import org.sonatype.nexus.repository.storage.AssetUpdatedEvent;
+import org.sonatype.nexus.repository.storage.Bucket;
+import org.sonatype.nexus.repository.storage.BucketStore;
+import org.sonatype.nexus.repository.storage.Component;
+import org.sonatype.nexus.repository.storage.ComponentDirector;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * @since 1.next
```
[the remark this answers is not in the document]
Comment (DreamVan) [reply nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RComponentDirector.java]: Done c78c8180544f6204b0063b1ba2c56a4abc81150b

**Pull request merged**